fw/mmap_buffer.h

```diff
@@ -0,0 +1,175 @@
+ *      variable-length records. Our use case requires handling multiple
+ *      records per page, making this ring buffer inefficient.
+ *
+ *  * tracefs ring buffer:
+ *      Like the packet_ring_buffer, this buffer is primarily designed for
+ *      page-level operations.
+ *
+ *  * BPF ring buffer
+ *      Involves sleepable functions, making it incompatible with the softirq
+ *      context. Also, bpf_ringbuf_reserve() requires the length of data to be
+ *      known in advance, which would force us to traverse the data twice.
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#ifndef __TFW_MMAP_BUFFER_H__
+#define __TFW_MMAP_BUFFER_H__
+
+#ifdef __KERNEL__
+
+#include "str.h"
+#include <linux/mm.h>
+#include <linux/fs.h>
+#include <linux/cdev.h>
+
+#else /* __KERNEL__ */
+
+#include <stdint.h>
+
+#define u32 uint32_t
+#define u64 uint64_t
+
+#endif /* __KERNEL__ */
+
+#define TFW_MMAP_BUFFER_DATA_OFFSET 32
+#define TFW_MMAP_BUFFER_MIN_SIZE    PAGE_SIZE
+#define TFW_MMAP_BUFFER_MAX_SIZE    (PAGE_SIZE * 4096)
+
+#define TFW_MMAP_BUFFER_MAX_NAME_LEN 32
+
+#define TFW_MMAP_BUFFER_DATA_SIZE(size) (size - TFW_MMAP_BUFFER_DATA_OFFSET)
+#define TFW_MMAP_BUFFER_FULL_SIZE(size) (size + TFW_MMAP_BUFFER_DATA_OFFSET)
+
+typedef struct {
+	u64 head;    /* head offset where the next data write will happen */
+	u64 tail;    /* tail offset where the next data read will happen */
+	u32 size;    /* size of the ring buffer data in bytes */
+	u32 cpu;     /* ID of the CPU tied to this buffer */
+#ifdef __KERNEL__
+	/* is_ready indicates that the buffer is mapped to user space and ready both
+	 * for writing and reading. Resetting this field signals to user space that it
+	 * should stop reading, unmap and close the file.
+	 */
+	atomic_t is_ready;
+#else
+	int is_ready;
+#endif
+	char __attribute__((aligned(TFW_MMAP_BUFFER_DATA_OFFSET))) data[];
+} TfwMmapBuffer;
+
+#ifdef __KERNEL__
+
+typedef struct {
+	TfwMmapBuffer __percpu **buf;
+	char dev_name[TFW_MMAP_BUFFER_MAX_NAME_LEN];
+	atomic_t dev_is_opened;
+	/* is_freeing indicates that freeing process started. It's necessary to
+	 * exclude repeated file opening.
+	 */
+	atomic_t is_freeing;
+	int dev_major;
+	struct class *dev_class;
+	struct page *pg[];
+} TfwMmapBufferHolder;
```

Comment:
Coding style: could you please align the data structure members and adjust comments like for example in https://github.com/tempesta-tech/tempesta/blob/master/fw/cache.c#L129 ?

Reply:
Fixed.